File: frontend/src/types/surgicalReport.ts
/**
 * Interface สำหรับรูปภาพ Microscopic ที่อยู่ใน Snapshot ของ Report
 */
export interface MicroscopicImageSnapshot {
  id: number;
  image_url: string;
  magnification?: string;
  stain?: string;
  description?: string;
}

/**
 * Interface สำหรับผู้ลงนามในรายงาน (Snapshot/Live Status)
 * เชื่อมโยงกับตาราง report_signers
 */
export interface ReportSignerResponse {
  id: number;
  report_id: number;
  user_id: number;
  role: "primary" | "co-signer" | "resident" | "consultant";
  diagnosis_order?: number;

  // สถานะการลงนาม
  signed_at: string | null; // ถ้าเป็น null แปลว่ายังไม่ได้กดเซ็น (Draft/Pending)
  assigned_at: string;

  // คำถาม/บริบทที่ primary ส่งมาให้ co-signer
  consult_note?: string | null;

  // ความเห็นของ co-signer
  agreement?: "agree" | "disagree" | null;
  agreement_note?: string | null;

  // ข้อมูล User (Join มาจากตาราง User)
  full_name?: string;
  user_full_name?: string;
  signature_path?: string;
}

/**
 * Interface หลักสำหรับ SurgicalReport (Snapshot)
 * อ้างอิงตาม SQLAlchemy Model: SurgicalReport
 */
export interface SurgicalReport {
  id: number;
  case_id: number;
  version_no: number;

  // --- ข้อมูลคนไข้ (Immutable Snapshot) ---
  accession_no: string; // เลขเบอร์เคส (S26-xxxx)
  patient_title?: string; // คำนำหน้า
  patient_name: string; // ชื่อ
  patient_ln: string; // นามสกุล (ln ใน model)
  patient_hn: string; // HN
  patient_cid: string; // เลขบัตรประชาชน
  patient_birth_date?: string; // ISO Date string
  patient_age_display?: string;
  patient_age: number; // อายุ ณ วันที่ออกผล
  patient_gender: string; // เพศ

  has_malignancy: boolean; // 🚩 เพิ่มตัวนี้
  has_critical: boolean; // 🚩 เพิ่มตัวนี้
  is_print: boolean;

  // --- ข้อมูลการรับสิ่งส่งตรวจ (Snapshot) ---
  collect_at?: string; // วันที่เก็บสิ่งส่งตรวจ
  registered_at?: string; // วันที่ลงทะเบียนรับเคส

  // --- แหล่งที่มาและผู้ส่ง (Snapshot) ---
  hospital_name?: string; // โรงพยาบาลต้นทาง
  department_name?: string; // แผนกที่ส่ง
  clinician_name?: string; // แพทย์เจ้าของไข้

  // --- เนื้อหารายงาน (Snapshot) ---
  clinical_history_snapshot?: string; // ประวัติทางคลินิก
  specimen_summary?: string; // รายการชิ้นเนื้อทั้งหมด
  gross_description_summary?: string; // รายละเอียดการตรวจด้วยตาเปล่า
  diagnosis_summary?: string; // ผลการวินิจฉัย
  microscopic_summary?: string; // รายละเอียดกล้องจุลทรรศน์
  comment_summary?: string; // ความเห็นเพิ่มเติม

  // --- ผู้ลงนามและประเภท ---
  report_type: "Final" | "Addendum" | "Correction";
  pathologist_name?: string; // ชื่อพยาธิแพทย์ ณ วันที่ลงนาม
  approved_by_id?: number;
  approver_name_snapshot?: string;

  // --- สถานะและเวลา ---
  status:
    | "draft"
    | "pending"
    | "published"
    | "cancelled"
    | "signed out"
    | "pending_approval";
  published_at?: string; // วันที่เผยแพร่ผล
  created_at: string;
  updated_at: string;

  // --- ความสัมพันธ์เพิ่มเติม (ถ้า Backend ส่งมา) ---
  gross_images?: MicroscopicImageSnapshot[];
  microscopic_images?: MicroscopicImageSnapshot[];
  approval_logs?: ApprovalLogResponse[];
  signers?: ReportSignerResponse[];
}

/**
 * Interface สำหรับการทำ Pagination
 */
export interface SurgicalReportPagination {
  items: SurgicalReport[];
  total: number;
  page: number;
  size: number;
}

/**
 * Interface สำหรับประวัติการอนุมัติ (Approval Log)
 */
export interface ApprovalLogResponse {
  id: number;
  report_id: number;
  approver_id: number;
  approver_name: string;
  action: "APPROVED" | "REJECTED" | "REQUEST_CHANGES";
  comment?: string;
  created_at: string;
}

/**
 * Interface สำหรับส่งข้อมูลการอนุมัติไปยัง Backend
 */
export interface ReportApproveRequest {
  action: "APPROVED" | "REJECTED" | "REQUEST_CHANGES";
  comment?: string;
  agreement?: "agree" | "disagree";
  agreement_note?: string;
}

/**
 * Interface สำหรับสถิติของ Surgical Cases
 */
export interface SurgicalStatResponse {
  total_cases: number;
  average_tt_days: number;
  average_tt_hours: number;
  daily_stats: {
    date: string;
    total_cases: number;
    average_tt_hours: number;
  }[];
  tt_distribution?: {
    tt_days: string;
    case_count: number;
  }[];
  complexity_breakdown?: {
    small: number;
    medium: number;
    large: number;
  };
}

export interface LabTechStatResponse {
  grossed_cases: number;
  embedded_blocks: number;
  sectioned_blocks: number;
  stained_blocks: number;
  total_slides: number;
  dispatched_cases: number;
  outlab_sent_blocks: number;
  complexity_breakdown?: {
    small: number;
    medium: number;
    large: number;
  };
}

File: backend/app/schemas/surgical_report.py
from pydantic import ConfigDict, BaseModel
from datetime import datetime, date
from typing import Optional, List


# Schema พื้นฐานสำหรับข้อมูล Snapshot
class SurgicalReportBase(BaseModel):
    case_id: Optional[int] = None
    version_no: int = 1
    report_type: str = "Final"
    status: str = "draft"

    # --- Patient Identifiers (Snapshot) ---
    accession_no: Optional[str] = None
    patient_title: Optional[str] = None
    patient_name: Optional[str] = None
    patient_ln: Optional[str] = None
    patient_hn: Optional[str] = None
    patient_cid: Optional[str] = None
    patient_birth_date: Optional[date] = None
    patient_age: Optional[int] = None
    patient_age_display: Optional[str] = None
    patient_gender: Optional[str] = None

    # --- Clinical Context ---
    hospital_name: Optional[str] = None
    department_name: Optional[str] = None
    clinician_name: Optional[str] = None
    clinical_history_snapshot: Optional[str] = None

    # --- Medical Content ---
    specimen_summary: Optional[str] = None
    gross_description_summary: Optional[str] = None
    diagnosis_summary: Optional[str] = None
    microscopic_summary: Optional[str] = None
    comment_summary: Optional[str] = None
    pathologist_name: Optional[str] = None

    # --- 🚩 เพิ่ม Identity & Footer Snapshots (เพื่อให้ Response มีข้อมูลหัว/ท้ายรายงาน) ---
    lab_name_th_snapshot: Optional[str] = None
    lab_name_en_snapshot: Optional[str] = None
    lab_address_snapshot: Optional[str] = None
    report_footer_snapshot: Optional[str] = None

    # --- Timestamps ---
    collect_at: Optional[datetime] = None
    registered_at: Optional[datetime] = None
    reported_at: Optional[datetime] = None  # 🚩 เพิ่มวันที่พยาธิแพทย์ส่งข้อมูล


class SurgicalReportCreate(SurgicalReportBase):
    # ใช้รับข้อมูลจาก prepare_report_data เพื่อ Insert ลง DB
    pass


class MicroscopicImageSnapshotSchema(BaseModel):
    id: int
    image_url: str
    magnification: Optional[str] = None
    stain: Optional[str] = None
    description: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ReportSignerCreate(BaseModel):
    user_id: int
    role: str = "co-signer"
    consult_note: Optional[str] = None


class ReportSignerResponse(BaseModel):
    user_id: Optional[int] = None
    role: str
    diagnosis_order: int = 1
    assigned_at: datetime
    signed_at: Optional[datetime] = None
    consult_note: Optional[str] = None
    agreement: Optional[str] = None
    agreement_note: Optional[str] = None
    user_full_name: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class SurgicalReportResponse(SurgicalReportBase):
    id: int
    published_at: Optional[datetime] = None  # วันที่ผลสมบูรณ์ (Public)
    is_print: bool = False
    approved_at: Optional[datetime] = None  # วันที่ได้รับการอนุมัติ (ถ้ามี)
    approved_by_id: Optional[int] = None
    approver_name_snapshot: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    signers: List[ReportSignerResponse] = []

    microscopic_images: List[MicroscopicImageSnapshotSchema] = []

    model_config = ConfigDict(from_attributes=True)


class SurgicalReportPagination(BaseModel):
    items: List[SurgicalReportResponse]
    total: int
    page: int
    size: int

    model_config = ConfigDict(from_attributes=True)


class DailyStat(BaseModel):
    date: str
    total_cases: int
    average_tt_hours: float

class TATDistribution(BaseModel):
    tt_days: str
    case_count: int

class SurgicalStatResponse(BaseModel):
    total_cases: int
    average_tt_days: float
    average_tt_hours: float
    daily_stats: List[DailyStat] = []
    tt_distribution: List[TATDistribution] = []
    complexity_breakdown: Optional[dict] = None


class LabTechStatResponse(BaseModel):
    grossed_cases: int = 0
    embedded_blocks: int = 0
    sectioned_blocks: int = 0
    stained_blocks: int = 0
    total_slides: int = 0
    dispatched_cases: int = 0
    outlab_sent_blocks: int = 0
    complexity_breakdown: Optional[dict] = None



